### PR TITLE
layout: Fix abspos position when it would be preceded by phantom line

### LINKED ISSUE #41990: Wrong abspos position when followed by a block, inside an inline FC
```html
<!DOCTYPE html>
<style>
#reference-overlapped-red {
  position: absolute;
  background-color: red;
  width: 100px;
  height: 100px;
  z-index: -1;
}
.test-inline-block-overlapping-green {
  display: inline-block;
  background-color: green;
  width: 100px;
  height: 100px;
}
</style>
<p>Test passes if there is a filled green square and <strong>no red</strong>.</p>
<span>
  <div id="reference-overlapped-red"></div>
  <div class="test-inline-block-overlapping-green"></div>
</span>
```

| Servo | Expected |
| - | - |
| Image | Image |

### PR DESCRIPTION
**Pull request by Loirooriol**
The static position of an abspos is the hypothetical position that it would have had in the normal flow. If the abspos is inside inline layout and it had a block-level original display, then the static position is computed as if the abspos was a block-level that breaks the inline.

Usually this places the static position after the current line. However, if the abspos is not preceded by any text, padding, border, margin, preserved newline or atomic inline, then the static position should be as if the abspos was preceded by a 0px tall phantom line. Previously we weren't taking that into account.

Note that browsers aren't completely interoperable in corner cases.

Testing: Adding 3 new tests. Blink fails the 2nd one and Gecko fails the 3rd one. WebKit and Servo pass them all.
Fixes: #41990

### CODE REVIEW

File: components/layout/flow/inline/line.rs [hunk 1]
```diff
@@ -673,9 +673,15 @@ impl LineItemLayout<'_, '_> {
...
+                // Note that phantom lines are treated as being zero-height for this purpose.
+                // <https://drafts.csswg.org/css-inline-3/#invisible-line-boxes>
                 LogicalVec2 {
                     inline: -self.current_state.parent_offset.inline,
-                    block: block_position + self.line_metrics.block_size,
+                    block: if absolute.on_phantom_line {
+                        block_position
+                    } else {
+                        block_position + self.line_metrics.block_size
+                    },
```
Comment (mrobinson) [inline components/layout/flow/inline/line.rs]: Is there a reason you can't check `self.is_phantom_line` here? Is it possible for a absolute to be listed as `on_phantom_line`, but not be taking part in a `LineItemLayout` for which `LineItemLayout::is_phantom_line` isn't true?

Comment (Loirooriol) [reply components/layout/flow/inline/line.rs]: The problem is that `self.is_phantom_line` refers to the final line, so it may be false because of contents after the abspos. But here we need the static position, which depends on whether the preceding contents would produce a phantom line or not. Because in the static position, the abspos behaves as in-flow, and thus it breaks the line.

Comment (Loirooriol) [reply components/layout/flow/inline/line.rs]: I have renamed and added a rustdoc comment for clarity

File: components/layout/flow/inline/line.rs [hunk 2]
```diff
@@ -936,6 +942,10 @@ impl AtomicLineItem {
 
 pub(super) struct AbsolutelyPositionedLineItem {
     pub absolutely_positioned_box: ArcRefCell<AbsolutelyPositionedBox>,
+    /// Whether the line would be phantom if it were to end before the abspos.
+    /// This is used when computign the static position (in the block axis) of
```
Comment (mrobinson) [inline components/layout/flow/inline/line.rs]:
```suggestion
    /// This is used when computing the static position (in the block axis) of
```

File: components/layout/flow/inline/line.rs [hunk 3]
```diff
@@ -936,6 +942,10 @@ impl AtomicLineItem {
...
+    /// an abspos whose original display had a block outer display type.
+    pub line_of_preceding_contents_would_be_phantom: bool,
```
Comment (mrobinson) [inline components/layout/flow/inline/line.rs]:
```suggestion
    pub preceding_line_content_would_produce_phantom_line: bool,
```

**Review by mrobinson — APPROVED** [review]
Looks good with two nits:
